src/graph.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of dot-jaeger.
+
+// dot-jaeger is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// dot-jaeger is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with dot-jaeger.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::primitives::{Span, TraceObject};
+use anyhow::Error;
+use daggy::{Dag, NodeIndex, Walker};
+use std::collections::HashMap;
+
+const EDGE_WEIGHT: u32 = 1;
+type DirectedGraph<'a> = Dag<Span<'a>, u32, u32>;
+
+#[derive(Debug)]
+pub struct Graph<'a> {
+	trace: &'a TraceObject<'a>,
+	graph: DirectedGraph<'a>,
+	/// Dictionary of the nodes present in the graph
+	index_lookup: HashMap<&'a str, NodeIndex<u32>>,
+	span_lookup: HashMap<NodeIndex, &'a str>,
+}
+
+impl<'a> Graph<'a> {
+	pub fn new(trace: &'a TraceObject<'a>) -> Result<Self, Error> {
+		let mut graph = Dag::new();
+		let mut index_lookup = HashMap::new();
+		let mut span_lookup = HashMap::new();
+
+		for span in trace.spans.values() {
+			let index = graph.add_node(span.clone());
+			index_lookup.insert(span.span_id, index);
+			span_lookup.insert(index, span.span_id);
+		}
+
+		for id in span_lookup.values() {
```

Comment:
Couldn't you do this work in the first for-loop? Maybe this is not perf sensitive and like this it's very readable, so up to you.

Reply:
yeah this could be done in one iteration, but it's not extremely performance sensitive and the speedup from using the library makes up for it anyway


with a DAG there's no checking for duplicate items, which would mean we'd have to have some way to make sure some span hasn't already been inserted if we don't separate adding nodes and then creating the edges for them. The JSON from jaeger also isn't necessarily in order either, so we'd have to deal with checking if a span exists in our lookup table, and t hen checking if it exists yet in the graph (which might require a search in itself)